cpp/inc/bond/ext/grpc/wait_callback.h

```diff
@@ -0,0 +1,175 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#pragma once
+
+#include <bond/core/bonded.h>
+#include <grpc++/impl/codegen/status.h>
+
+#include <boost/optional.hpp>
+
+#include <chrono>
+#include <condition_variable>
+#include <functional>
+#include <memory>
+#include <mutex>
+
+namespace bond { namespace ext { namespace gRPC {
+
+/// @brief A callback type that can be manually waited upon.
+///
+/// The \ref wait member function can be used to wait until the callback
+/// has been called. Then, the \ref status and \ref response member
+/// functions can be called to inspect the results.
+///
+/// @example wait_callback.cpp
+template <typename TResponse>
+class wait_callback
+{
+public:
+    wait_callback() : _impl(std::make_shared<impl>()) { }
+
+    using CallbackType = std::function<void(const bond::bonded<TResponse>&, const grpc::Status&)>;
+
+    /// @brief Creates a callback that will signal this wait_callback
+    /// instance.
+    ///
+    /// If multiple callbacks are created and invoked, only the results of
+    /// the first one that is invoked will be recorded.
+    ///
+    /// It is safe for callbacks to outlive the wait_callback that created
+    /// them. When they are invoked their results will be discarded.
+    CallbackType callback()
```

Comment:
It would probably be better to return the lambda itself (auto). The caller can always decide to store it in std::function, but it won't be enforced by this implementation.